File: Compiler/Ownership/MemberInfo.py
import Compiler.DependencyProcessor as DependencyProcessor

class MemberKind(object):
    def __init__(self):
        self.type = None
        self.index = 0

    
class MemberInfo(object):
    def __init__(self):
        self.root = None
        self.kind = None
        self.info = None

    def __str__(self):
        return "%s/%s.%s/%s" % (self.kind.type, self.root, self.kind.index, self.info)

    def __repr__(self) -> str:
        return self.__str__()

def getGroupDependencyMap(members):
    dep_map = {}
    for member_info in members:
        dep_map[member_info.root] = []
        dep_map[member_info.info.group_var] = []
    for member_info in members:
        dep_map[member_info.root].append(member_info.info.group_var)
    return dep_map

def calculateChildOwnershipVars(members):
    child_ownership_vars = {}
    for member in members:
         child_ownership_vars[member.root] = []
    for member in members:
         child_ownership_vars[member.root].append(member.info.ownership_var)
    return child_ownership_vars

def collectDepOwnershipVarsForGroupVar(group, deps_map, ownership_dep_map, ownership_vars, item):
    deps = deps_map[item]
    for dep in deps:
        if dep in ownership_dep_map:
            ownership_vars = ownership_vars + ownership_dep_map[dep]
    return ownership_vars

def calculateDepsForGroup(child_ownership_vars, ownership_dep_map, deps_map, group):
    ownership_vars = []
    for item in group:
        if item in child_ownership_vars:
            ownership_vars += child_ownership_vars[item]
            ownership_vars = collectDepOwnershipVarsForGroupVar(group, deps_map, ownership_dep_map, ownership_vars, item)
    ownership_vars = list(set(ownership_vars))
    for item in group:
        ownership_dep_map[item] = ownership_vars
    return ownership_dep_map

def calculateOwnershipDepMap(members):
    deps_map = getGroupDependencyMap(members)
    groups = DependencyProcessor.processDependencies(deps_map)
    child_ownership_vars = calculateChildOwnershipVars(members)
    ownership_dep_map = {}
    for group in groups:
        ownership_dep_map = calculateDepsForGroup(child_ownership_vars, ownership_dep_map, deps_map, group.items)
    return ownership_dep_map

File: Siko/src/siko/ownership/MemberInfo.rs
use super::TypeVariableInfo::{GroupTypeVariable, OwnershipTypeVariable, TypeVariableInfo};
use crate::siko::util::DependencyProcessor;
use std::{collections::BTreeMap, fmt::Display};

#[derive(Debug, PartialEq, Eq, PartialOrd, Ord)]
pub enum MemberKind {
    Variant,
    Field,
    Extern,
}

impl Display for MemberKind {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match &self {
            MemberKind::Variant => write!(f, "v"),
            MemberKind::Field => write!(f, "f"),
            MemberKind::Extern => write!(f, "e"),
        }
    }
}

pub struct MemberInfo {
    pub root: GroupTypeVariable,
    pub kind: MemberKind,
    pub index: u32,
    pub info: TypeVariableInfo,
}

impl Display for MemberInfo {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "{}/{}/{}/{}",
            self.kind, self.index, self.root, self.info
        )
    }
}

fn getGroupDependencyMap(
    members: &Vec<MemberInfo>,
) -> BTreeMap<GroupTypeVariable, Vec<GroupTypeVariable>> {
    let mut depMap = BTreeMap::new();
    for member in members {
        depMap.insert(member.root, Vec::new());
        depMap.insert(member.info.group, Vec::new());
    }
    for member in members {
        depMap
            .get_mut(&member.root)
            .unwrap()
            .push(member.info.group);
    }
    depMap
}

fn calculateChildOwnershipVars(
    members: &Vec<MemberInfo>,
) -> BTreeMap<GroupTypeVariable, Vec<OwnershipTypeVariable>> {
    let mut childOwnershipMap = BTreeMap::new();
    for member in members {
        childOwnershipMap.insert(member.root, Vec::new());
    }
    for member in members {
        childOwnershipMap
            .get_mut(&member.root)
            .unwrap()
            .push(member.info.owner);
    }
    childOwnershipMap
}

fn collectDepOwnershipVarsForGroupVar(
    depMap: &BTreeMap<GroupTypeVariable, Vec<GroupTypeVariable>>,
    ownership_dep_map: &BTreeMap<GroupTypeVariable, Vec<OwnershipTypeVariable>>,
    mut ownershipVars: Vec<OwnershipTypeVariable>,
    item: GroupTypeVariable,
) -> Vec<OwnershipTypeVariable> {
    let deps = depMap.get(&item).unwrap();
    for dep in deps {
        if let Some(vars) = ownership_dep_map.get(dep) {
            ownershipVars.extend(vars);
        }
    }
    return ownershipVars;
}

fn calculateDepsForGroup(
    childOwnershipVars: &BTreeMap<GroupTypeVariable, Vec<OwnershipTypeVariable>>,
    mut ownershipDepMap: BTreeMap<GroupTypeVariable, Vec<OwnershipTypeVariable>>,
    depMap: &BTreeMap<GroupTypeVariable, Vec<GroupTypeVariable>>,
    group: &Vec<GroupTypeVariable>,
) -> BTreeMap<GroupTypeVariable, Vec<OwnershipTypeVariable>> {
    let mut ownershipVars = Vec::new();
    for item in group {
        if let Some(vars) = childOwnershipVars.get(&item) {
            ownershipVars.extend(vars);
            ownershipVars =
                collectDepOwnershipVarsForGroupVar(&depMap, &ownershipDepMap, ownershipVars, *item);
        }
    }
    ownershipVars.sort();
    ownershipVars.dedup();
    for item in group {
        ownershipDepMap.insert(*item, ownershipVars.clone());
    }
    ownershipDepMap
}

pub fn calculateOwnershipDepMap(
    members: &Vec<MemberInfo>,
) -> BTreeMap<GroupTypeVariable, Vec<OwnershipTypeVariable>> {
    let depMap = getGroupDependencyMap(members);
    let groups = DependencyProcessor::processDependencies(&depMap);
    let childOwnershipVars = calculateChildOwnershipVars(members);
    let mut ownershipDepMap = BTreeMap::new();
    for group in groups {
        ownershipDepMap =
            calculateDepsForGroup(&childOwnershipVars, ownershipDepMap, &depMap, &group.items);
    }
    return ownershipDepMap;
}
